### PR TITLE
Better "New Contributor" Pagination

### LINKED ISSUE #3556: "Add Contributors" modal needs pagination
## Steps

1. Log in to osf.io.
2. Navigate to a project for which you have admin access
3. Go to Sharing tab
4. Click to Add Contributors
5. Add more then ten contributors
6. Now navigate to a component within the project for which you are the only contributor
7. Go to the Sharing tab for the component
8. Click to Add Contributors
9. Click "Import contributors from [Insert project name]." This will import contributors from the parent project. 
10. Notice the contributors list.

## Expected Behavior

I expect pagination to organize results when there are many contributors imported.

## Actual Behavior

No pagination occurs even though there are 11 contributors listed. On a larger project, there would be an even longer list of contributors there. I expect pagination to organize results as in the "Add links" modal (screenshot below).

## Screenshot

**Add contributors modal no pagination**

no_paginator

**Add links modal pagination**

add_links_modal_pagination

### PR DESCRIPTION
**Pull request by Johnetordoff**
<h2>Purpose</h2>
Add pagination to to "Add Contributor" modal when contributors are imported from the parent (closes #3556) and fixes bug with persisting paginatior buttons when user selects "Most Common Projects" or "Most Recent Collaborators" as contributor searches.
<h2>Changes</h2>
makes all search use paginator fetchResults() and adds function that collates results for paginator pages.
<h2>Side Effects</h2>
None that I know of.

### CODE REVIEW

**Comment by chennan47** [comment]
@Johnetordoff all the pagination calculation logic should be on server side. You can just modify the url by make a new observable, this.url, change this url valuable in different method before the fetchResult is called.

**Comment by Johnetordoff** [comment]
It's a little more complicated then that because search_contributor is the only search that actually makes an elastic search query, so it the only one that has sever-side pagination calculation already, that's why they weren't originally paginated, I could go add individual pagination calculation to all them, but obviously a single function is easier.

Also why should the url be an observable? It's only be used internally, right?

**Comment by sloria** [comment]
@Johnetordoff Getting this up to date with develop should fix the build. Can you do that?

**Comment by Johnetordoff** [comment]
@sloria its up to date now

File: website/util/permissions.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 # NOTE: Ordered from most-restrictive to most permissive
 PERMISSIONS = [READ, WRITE, ADMIN]
 CREATOR_PERMISSIONS = [READ, WRITE, ADMIN]
-DEFAULT_CONTRIBUTOR_PERMISSIONS = [READ, WRITE]
+DEFAULT_CONTRIBUTOR_PERMISSIONS = [READ, WRITE, ADMIN]
```
Comment (chennan47) [inline website/util/permissions.py]: I don't think this is the change included in this pr. Please remove this change if it's not what you mean to do.

File: website/project/views/contributor.py [hunk 2]
```diff
@@ -135,7 +136,18 @@ def get_contributors_from_parent(auth, node, **kwargs):
...
+
+    contribs, pages = paginate(contribs,
+                               len(contribs),
+                               page,
+                               size)
+    contribs = list(contribs)
+
+    return {'users': contribs,
+            'page': page,
+            'pages': pages}
```
Comment (chennan47) [inline website/project/views/contributor.py]: This code block is repeated 3 times. It is worthy to write a help function to reduce duplicate code.

File: website/static/js/contribAdder.js [hunk 3]
```diff
@@ -112,119 +113,61 @@ var AddContributorViewModel = oop.extend(Paginator, {
...
-        }
-    },
-    importFromParent: function() {
-        var self = this;
-        self.notification(false);
-        $.getJSON(
-            nodeApiUrl + 'get_contributors_from_parent/', {},
-            function(result) {
-                if (!result.contributors.length) {
+                    if (!contributors.length) {
```
Comment (chennan47) [inline website/static/js/contribAdder.js]: This check should go before the new contributor constructor.

File: website/static/js/contribAdder.js [hunk 4]
```diff
@@ -112,119 +113,61 @@ var AddContributorViewModel = oop.extend(Paginator, {
...
         self.notification(false);
-        if (self.query()) {
             return $.getJSON(
-                '/api/v1/user/search/', {
-                    query: self.query(),
-                    excludeNode: nodeId,
-                    page: self.pageToGet
-                },
+                self.url() + '?size=' + RESULTS_PER_PAGE + '&page='+ this.pageToGet(),
+                this.searchParams,
```
Comment (chennan47) [inline website/static/js/contribAdder.js]: Where do you initiate this variable?

File: website/views.py [hunk 5]
```diff
@@ -244,7 +244,11 @@ def validate_page_num(page, pages):
 
 
 def paginate(items, total, page, size):
-    pages = math.ceil(total / float(size))
+    try:
+        pages = math.ceil(total / float(size))
```
Comment (chennan47) [inline website/views.py]: I don't think this try and except is the right way to go. You should instead do a check if size is 0 then add some code to handle that specific situation.

**Comment by chennan47** [comment]
@Johnetordoff first pass done.

File: website/search/elastic_search.py [hunk 6]
```diff
@@ -440,6 +441,9 @@ def search_contributor(query, page=0, size=10, exclude=None, current_user=None):
         most recent employment and education, gravatar URL of an OSF user
 
     """
+    page = int(request.args.get('page', 0))
```
Comment (sloria) [inline website/search/elastic_search.py]: What if the client passes an invalid value, e.g. a string that isn't an integer? Handle errors correctly here.

**Comment by sloria** [comment]
Just a few driveby comments; I haven't fully reviewed this. Ping @chennan47 for second pass.

File: website/project/views/contributor.py [hunk 7]
```diff
@@ -732,3 +733,17 @@ def claim_user_post(node, **kwargs):
         'email': email,
         'fullname': unclaimed_data['name']
     }
+
+def paginateContributorSearch(contribs, **kwargs):
+    size = int(request.args.get('size', len(contribs)))
+    page = int(request.args.get('page', 0))
```
Comment (chennan47) [inline website/project/views/contributor.py]: as @sloria said, you need error handler here.

File: website/static/js/contribAdder.js [hunk 8]
```diff
@@ -59,6 +60,7 @@ var AddContributorViewModel = oop.extend(Paginator, {
         self.results = ko.observableArray([]);
         self.selection = ko.observableArray();
         self.notification = ko.observable('');
+        self.url = ko.observable('');
```
Comment (chennan47) [inline website/static/js/contribAdder.js]: This is in order to pass in different url for ajax request based on different search options.

File: website/views.py [hunk 9]
```diff
@@ -244,7 +244,12 @@ def validate_page_num(page, pages):
 
 
 def paginate(items, total, page, size):
-    pages = math.ceil(total / float(size))
+
+    if size != 0:
```
Comment (chennan47) [inline website/views.py]: When there is no result return?

File: website/static/js/contribAdder.js [hunk 10]
```diff
@@ -14,7 +14,7 @@ var Paginator = require('./paginator');
 
 var NODE_OFFSET = 25;
 // Max number of recent/common contributors to show
-var MAX_RECENT = 5;
+var RESULTS_PER_PAGE = 7;
```
Comment (chennan47) [inline website/static/js/contribAdder.js]: I will prefer it t be 5 for consistency.

**Comment by chennan47** [comment]
@Johnetordoff @sloria second pass done. Some minor changes.

File: website/views.py (hunk 9 above)
Comment (Johnetordoff) [inline website/views.py]: I included this because it was failing tests

**Comment by sloria** [comment]
The New Contributor UI has diverged significantly since this PR was made. Closing this for the time being.

**Pull request closed without merging**